PermutationGroup.is_normal: Fixed check for identity group, revised tests

FabianBall opened the pull request:
#### References to other Issues or PRs
This fixes #15636 
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
The implementation that uses PermutationGroup.coset_factor to check if a
permutation p = g*h*g^{-1} (g \in G, H \in H, H is tested to be a normal
subgroup of G) is again in H (that is the PermutationGroup instance on
which the method is called) has the problem that if h is the identity
permutation the check fails. As a consequence, the result of the
is_normal check is incorrect. This contribution extends to check to
recognize the identity permutation.
Furthermore, the subgroup property was never checked which also led to
incorrect results (i.e. a group of larger order was identified to be a
normal subgroup of a group of lower order).
Tests are extended and an incorrect test is removed.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* combinatorics
  * `PermutationGroup.is_normal` did not perform a check if the group that is tested to be a normal subgroup is even a subgroup.
  * When the identity group (a `PermutationGroup` that is generated by the identity permutation) was tested to be a normal subgroup of any other group, `PermutationGroup.is_normal` gave the incorrect result (`False` instead of `True`).

<!-- END RELEASE NOTES -->

jksuom commented on sympy/combinatorics/perm_groups.py:
```diff
@@ -1830,6 +1832,8 @@ def is_normal(self, gr, strict=True):
         for g1 in gens1:
             for g2 in gens2:
                 p = _af_rmuln(g1, g2, _af_invert(g1))
+                if all(i == p[i] for i in range(len(p))):
```
There is  a method, `.is_Identity` for testing this. Moreover, it would be more efficient to test `g1`. It that is the identity, then `p` would be the identity for each `g2` and there is no need to loop over `gens2`.

FabianBall replied:
I'm aware of the `.is_Identity` method, which is testing exactly what I provided. However, I found it more efficient to not first create a `Permutation` instance.
But you are right with the 2nd concern, there even is an `is_trivial` property for `PermutationGroup`s, which is definitely a better solution.

jksuom replied:
Is this identity test included somewhere?

FabianBall replied:
No, it is not, but it is not necessary. In ``PermutationGroup.__new__`` there is a check that removes all identity permutations if there is more than one argument/permutation given. Therefore, either only non-trivial permutations remain or there is only one argument that is the identity permutation. In the second case, ``is_trivial`` does its job.
This means it is impossible to create a ``PermutationGroup`` instance that contains the identity permutation in the list of generators other than the identity group.

However, there is another flaw, but it is not directly related to this issue: There is a somewhat undocumented ``dups=False`` kwarg which disables duplicate removal.
``PermutationGroup(Permutation(5), Permutation(5), dups=False)`` results in an error, but not related to normal subgroup testing.

jksuom commented:
I'm not sure if a non-trivial group might have the identity among its generators. In that case, it seems that `g1` would still have to be checked.

FabianBall commented:
Ouch, I'm getting more and more into it and it's not getting easier... From what I have seen (and tested), the identity is removed from the list of permutations/generators when a new `PermutationGroup` is instantiated. The only possibility for the identity to be part of the generating set is if the group _is_ the trivial group. However, I would suggest not to rely on how this is implemented at the moment (though the behavior makes sense). This is also the reason why I chose the less efficient implementation at first.
What I found out just is that the behavior between `is_subgroup` and `is_normal` differs: the first performs type checking and only allows the (possible) subgroup to act on less elements (but not the other way round.
This is of course not directly related, but it is nonetheless related as in my opinion the behavior for both should be similar.
For now I suggest the following order of checks within `is_normal` (G is `self`, G' is the other group):
1. Check if G is a subgroup of G'. If not, G can't be a normal subgroup.
2. Check if G is the trivial group and if either the degrees of G and G' are the same or if we do not want to be strict. If yes, G is a normal subgroup and no further checks are needed.
3. Test strictness as before
4. Test the actual property of a normal subgroup as before.

oscarbenjamin commented:
It looks as if the tests have failed because of an unrelated issue. If you merge in (or rebase) master they should pass.

jksuom commented:
100 commits doesn't look right. A merge issue?

FabianBall commented:
I rebased sympy/master into my branch as @oscarbenjamin pointed out. Has it something to do with that?
I'm no git(hub) pro, therefore I'm not really sure. I hope I did not mess anything up?!

oscarbenjamin commented:
Before trying to fix anything make sure you've saved the work that you've done somewhere outside of git.

Then you can fix this with something like:
```console
    $ git remote add upstream https://github.com/sympy/sympy.git
    $ git fetch upstream
    $ git log # Find a commit that is older than these 100 and that is definitely on upstream/master
    $ git log upstream/master
```

Assuming you've copied the hash for the commit as abcde then (careful with this - make sure your work is saved somewhere else before resetting):
```console
    $ git checkout 15636_perm_group_is_normal
    $ git reset --hard abcde
    $ git rebase upstream/master
```
Then this branch will be the same as upstream/master. If something goes wrong during the rebase you can cancel with `git rebase --abort` although that won't undo the reset.

Assuming you're now back in sync with master you can reapply the changes you made, commit then again and then push to this PR with
```
    $ git push --force
```

Commit:
PermutationGroup.is_normal: Fixed check for identity group, revised tests

The implementation that uses PermutationGroup.coset_factor to check if a
permutation p = g*h*g^{-1} (g \in G, H \in H, H is tested to be a normal
subgroup of G) is again in H (that is the PermutationGroup instance on
which the method is called) has the problem that if h is the identity
permutation the check fails. As a consequence, the result of the
is_normal check is incorrect. This contribution extends to check to
recognize the identity permutation.
Furthermore, the subgroup property was never checked which also led to
incorrect results (i.e. a group of larger order was identified to be a
normal subgroup of a group of lower order).
Tests are extended and an incorrect test is removed.

FabianBall commented:
Thank you very much for assisting and sorry for causing a bit of a trouble! I hope everything is well now.

oscarbenjamin commented:
It's no trouble. Thanks for fixing it!

jksuom commented on sympy/combinatorics/perm_groups.py:
```diff
@@ -1817,8 +1817,12 @@ def is_normal(self, gr, strict=True):
+            return False
         d_self = self.degree
         d_gr = gr.degree
+        if self.is_trivial and (d_self == d_gr or not strict):
```
It seems that if `self` and `gr` are both trivial, then `d_self == d_gr` is true and therefore `True` will be returned independently of the value of `strict`.

FabianBall replied:
I can't follow. If ``self`` and ``gr`` are trivial but ``d_self == d_gr``, then it does not matter if ``strict`` is true or false, as both groups already act on the same number of elements.
If ``d_self != d_gr`` and ``strict==True`` we continue as before.

Maybe there is a misunderstanding what ``strict`` does?

jksuom replied:
Ignore that. I was under the impression that `strict` would mean "strict subgroup".

jksuom commented:
Thanks, this looks good.
